### PR TITLE
Faster serialization

### PR DESCRIPTION
**Pull request by michelemottini**
New FhirJsonFastSerializer and FhirXmlFastSerialize - performance testing using PerfTest.exe (just added):

```
JSON serialize X 10,000: 17,463.0ms
JSON fast serialize X 10,000: 1,133.0ms
XML serialize X 10,000: 12,913.0ms
XML fast serialize X 10,000: 1,080.0ms
```

Also: `Model.Version` enumeration values are now flags, that allow some simplifications

@scottfavre

### CODE REVIEW

**Comment by michelemottini** [comment]
The package (rename it to .nupkg):

[CareEvolution.Fhir.Core.1.4.0.zip](https://github.com/CareEvolution/fhir-net-api/files/3882992/CareEvolution.Fhir.Core.1.4.0.zip)

File: src/Hl7.Fhir.Core/Model/Generate.csx [hunk 1]
```diff
@@ -2758,13 +2811,48 @@ public class PropertyDetails
...
+        var isRequired = CardMin == "1" ?
+            "true" :
+            "false";
+        var elementDescription = $"\"{FhirName}\", {elementVersions}, {summaryVersions}, {isRequired}";
+        if (!IsMultiCard())
+        {
+            var isChoice = AllowedTypesByVersion.Any(pair => pair.Value.Count > 1) ?
+                "true" :
+                "false";
+            yield return $"sink.Element({elementDescription}, {isChoice}); {Name}?.Serialize(sink);";
```
Comment (scottfavre) [inline src/Hl7.Fhir.Core/Model/Generate.csx:2827]: minor: why have these two statements on a single line? makes the generated code harder to read.

File: src/Hl7.Fhir.Core/Serialization/SerializerSink.cs [hunk 2]
```diff
@@ -0,0 +1,863 @@
...
+            public bool IsChoice { get; set; }
+
+            public string GetElementName(string type)
+            {
+                if (IsChoice) return Name + type.Substring(0, 1).ToUpperInvariant() + type.Substring(1);
+                return Name;
+            }
+        }
+
+        private const string ObservationValueSystem = "http://hl7.org/fhir/v3/ObservationValue";
```
Comment (scottfavre) [inline src/Hl7.Fhir.Core/Serialization/SerializerSink.cs]: is this URL the same across all versions? looks like it is for V3

**Review by scottfavre — APPROVED** [review]
Looks good. A couple of minor things.

I plugged the sample nuget into rosetta and loading a big page of value sets (500 value sets, 8.6 MB of json) dropped average response time from 6.4 sec to 1.6 sec.

**Comment by michelemottini** [comment]
> I plugged the sample nuget into rosetta and loading a big page of value sets (500 value sets, 8.6 MB of json) dropped average response time from 6.4 sec to 1.6 sec.

Good - were you able to check if the generated JSON is OK?

File: src/Hl7.Fhir.Core/Serialization/SerializerSink.cs (hunk 2 above)
Comment (michelemottini) [reply src/Hl7.Fhir.Core/Serialization/SerializerSink.cs]: That `v3` stands for HL7v3 - not FHIR STU3 - still maybe that system _did_ change in R4 (to `http://terminology.hl7.org/CodeSystem/v3-ObservationValue`) ... I am asking ....

File: src/Hl7.Fhir.Core/Model/Generate.csx (hunk 1 above)
Comment (michelemottini) [reply src/Hl7.Fhir.Core/Model/Generate.csx:2827]: because they make the code easier to read!

**Comment by scottfavre** [comment]
Yes, all rosetta smoke tests passed using the official Fhir client, so the
JSON for term stuff is valid. I did not check any of the other resources.

**Comment by michelemottini** [comment]
> Yes, all rosetta smoke tests passed

OK, excellent, thanks again @scottfavre 

(and yes, they did change that URL in R4 - to `http://terminology.hl7.org/CodeSystem/v3-ObservationValue` (obviously) - fixed)

**Pull request merged**